components/webcompat_reporter/browser/webcompat_report_uploader_unittest.cc

```diff
@@ -0,0 +1,155 @@
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/cpp/weak_wrapper_shared_url_loader_factory.h"
+#include "services/network/test/test_url_loader_factory.h"
+#include "third_party/googletest/src/googletest/include/gtest/gtest.h"
+
+namespace webcompat_reporter {
+
+class WebcompatReportUploaderUnitTest : public testing::Test {
+ public:
+  WebcompatReportUploaderUnitTest()
+      : shared_url_loader_factory_(
+            base::MakeRefCounted<network::WeakWrapperSharedURLLoaderFactory>(
+                &url_loader_factory_)),
+        task_environment_(base::test::TaskEnvironment::TimeSource::MOCK_TIME) {
+    webcompat_report_uploader_ =
+        std::make_unique<WebcompatReportUploader>(shared_url_loader_factory_);
+  }
+
+  using OnRequestPayloadCallback =
+      base::RepeatingCallback<void(const std::string& request_payload)>;
+
+  void SetInterceptor(
+      OnRequestPayloadCallback* request_payload_callback = nullptr) {
+    url_loader_factory_.SetInterceptor(base::BindLambdaForTesting(
+        [&, request_payload_callback](const network::ResourceRequest& request) {
+          url_loader_factory_.ClearResponses();
+          if (request_payload_callback) {
+            std::string request_string(request.request_body->elements()
+                                           ->at(0)
+                                           .As<network::DataElementBytes>()
+                                           .AsStringPiece());
+            request_payload_callback->Run(request_string);
+          }
+        }));
+  }
+
+  WebcompatReportUploader* GetWebcompatUploader() {
+    return webcompat_report_uploader_.get();
+  }
+
+  void TestReportGeneration(const Report& report, const std::string& content) {
+    auto content_vals = base::test::ParseJson(content);
+    EXPECT_TRUE(content_vals.is_dict());
+    auto& content_dict_vals = content_vals.GetDict();
+
+    base::RunLoop run_loop;
+    auto request_payload_check_callback =
+        base::BindLambdaForTesting([&](const std::string& request_payload) {
+          auto request_vals = base::test::ParseJson(request_payload);
+          EXPECT_TRUE(request_vals.is_dict());
+
+          auto& request_dict_vals = request_vals.GetDict();
+
+          for (base::Value::Dict::iterator it = content_dict_vals.begin();
+               it != content_dict_vals.end(); ++it) {
+            EXPECT_NE(request_dict_vals.Find(it->first), nullptr);
+            EXPECT_EQ(*(request_dict_vals.Find(it->first)), it->second);
+          }
+          run_loop.Quit();
+        });
+
+    SetInterceptor(&request_payload_check_callback);
+    GetWebcompatUploader()->SubmitReport(report);
+    run_loop.Run();
+  }
+
+ protected:
+  std::unique_ptr<WebcompatReportUploader> webcompat_report_uploader_;
+  network::TestURLLoaderFactory url_loader_factory_;
+  scoped_refptr<network::SharedURLLoaderFactory> shared_url_loader_factory_;
+  content::BrowserTaskEnvironment task_environment_;
```

Comment:
is `base::test::TaskEnvironment` not enough?

Reply:
we need BrowserThread for CreateAndStartURLLoader()

Reply:
> we need BrowserThread for CreateAndStartURLLoader()

you only need this for the `DCHECK`, right? if that's the case, then you should go for SequencedChecker instead. Do you really need to depend on `content` at all? Because if you do, then from the correctness perspective, you should have all these parts be in `//brave/components/webcompat_reporter/content/`.